lms/tasks/grading.py

```diff
@@ -0,0 +1,133 @@
+from sqlalchemy import exists, select
+
+from lms.models import (
+    ApplicationInstance,
+    GradingSync,
+    GradingSyncGrade,
+    Grouping,
+    LTIRegistration,
+)
+from lms.services import LTIAHTTPService
+from lms.services.lti_grading.factory import LTI13GradingService
+from lms.tasks.celery import app
+
+
+@app.task()
+def sync_grades():
+    """Start processing pending (scheduled) GradingSync."""
+    with app.request_context() as request:
+        with request.tm:
+            scheduled_syncs = request.db.scalars(
+                select(GradingSync)
+                .where(GradingSync.status == "scheduled")
+                # We'll call this task once per new GradingSync but if we find more pending ones also start those.
+                .limit(5)
+                .with_for_update()
+            )
+
+            for sync in scheduled_syncs:
+                assignment = sync.assignment
+                assert (
+                    assignment.lis_outcome_service_url
+                ), "Assignment without grading URL"
+                lti_registration = request.db.scalars(
+                    select(LTIRegistration)
+                    .join(ApplicationInstance)
+                    .join(Grouping)
+                    .where(Grouping.id == assignment.course_id)
+                    .order_by(LTIRegistration.updated.desc())
+                ).first()
+                assert lti_registration, "No LTI registraion for LTI1.3 assignment"
+
+                for grade in sync.grades:
+                    sync_grade.delay(
+                        lti_registration_id=lti_registration.id,
+                        lis_outcome_service_url=assignment.lis_outcome_service_url,
+                        grading_sync_grade_id=grade.id,
+                    )
+
+                sync.status = "in_progress"
+
+
+@app.task(
+    max_retries=2,
+    retry_backoff=10,
+    autoretry_for=(Exception,),
+)
+def sync_grade(
+    *, lti_registration_id, lis_outcome_service_url: str, grading_sync_grade_id: int
+):
+    """Send one particular grade to the LMS."""
+    with app.request_context() as request:
+        with request.tm:
+            grading_sync_grade = request.db.get(GradingSyncGrade, grading_sync_grade_id)
+            grading_sync = grading_sync_grade.grading_sync
+
+            grading_service = LTI13GradingService(
+                ltia_service=request.find_service(LTIAHTTPService),
+                line_item_url=None,
+                line_item_container_url=None,
+                product_family=None,  # type: ignore
+                misc_plugin=None,  # type: ignore
+                lti_registration=None,  # type: ignore
+            )
```

Comment:
Can you remind me why it was not possible to inject this instead of creating an instance on the fly?

Reply:
The service factory (well the whole service) is meant to be used in the context of a launch. 

We should refactor it so most of these are injected in method calls instead of class attributes. But unfortunately we'd need to do that for both LTI1.1 and LTI1.3 

[There's a note on the BE issue about this.](https://github.com/hypothesis/lms/issues/6647)

Reply:
Thanks!